modules/apps/asset/asset-taglib/src/main/resources/META-INF/resources/asset_categories_navigation/js/AssetCategoriesNavigationTreeView.js

```diff
@@ -15,13 +15,41 @@
 import {Treeview} from 'frontend-js-components-web';
 import React from 'react';
 
-const AssetCategoriesNavigationTreeView = ({categories, namespace}) => {
+function buildNodes(vocabularies, categories) {
+	const nodes = [];
+
+	vocabularies.forEach((vocabulary) => {
+		const child = {
+			...vocabulary,
+			children: [],
+			id: vocabulary.vocabularyId,
+		};
+
+		categories.forEach((category) => {
+			if (category.vocabularyId === child.id) {
+				child.children.push(category);
+			}
+		});
+
+		nodes.push(child);
+	});
+
+	return nodes;
+}
```

Comment:
@julien (Copy-pasting from https://github.com/mateomustapic/liferay-portal/pull/9):

---

In general a pattern like:

```js
const nodes = [];

vocabularies.forEach((_) => {
    // ...

    nodes.push(child);
});

return nodes;
```

Usually just wants to be:

```js
return vocabularies.map((_) => {
    // ...

    return child;
});
```

You know, functional/immutable generally preferable to imperative/mutable and all that. Depending on how far you want to go down that path, in this one, you have a nested `forEach` that has the same kind of pattern; could be:

```js
const child = {
	...vocabulary,
	children: categories
		.map((category) => {
			if (category.vocabularyId === child.id) {
				return {
					...category,
					id: category.categoryId,
				};
			}
		})
		.filter(Boolean),
};
```

(ie. using `filter(Boolean)` to skip empty `map` results)

All together, that would be:

```js
function buildNodes(vocabularies, categories) {
	return vocabularies.map((vocabulary) => ({
		...vocabulary,
		children: categories
			.map((category) => {
				if (category.vocabularyId === child.id) {
					return {
						...category,
						id: category.categoryId,
					};
				}
			})
			.filter(Boolean),
	}));
}
```

Having said that, are the multiple levels of nesting possible? Handling those would imply the need for a recursive solution (or equivalent).

Reply:
Also pasting from #9 :joy: 
@wincent, that's great, thanks for the help.
Yes, we will need something recursive because categories can also have nested children than can also have nested children, but I think that based on your feedback we can come up with something that works.

Reply:
Cool. Don't feel obliged to go too far with the function/immutable stuff. I just spelled it all out there for completeness. For me, "step 1" (ie. moving from `forEach()` + `push()` to just `map()`) is a clear win, but "step 2" (ie. moving from nested `forEach()` + `push()` to `map()` + `filter()`) is not as obvious. The functional style can be harder to read for the uninitiated, but I tend to lean towards it because I think it's something that we are all tending to get more and more used to over time. 🤷

Reply:
> Don't feel obliged to go too far with the function/immutable stuff. 

Trust me, I won't :joy_cat: 

But I still want to get this right

Reply:
> Yes, we will need something recursive because categories can also have nested children than can also have nested children, but I think that based on your feedback we can come up with something that works.

Well it seems that's wrong, I'm not able to add a category inside a category.
But I did updated the code

Reply:
I pushed [a commit](https://github.com/wincent/liferay-portal/pull/311/commits/bf40a02dd994673001df3439f5103633d6fcdfe2) to simplify [the final implementation](https://github.com/wincent/liferay-portal/pull/311/commits/ea5b588ee5c7e51ee98acf1182cbe73899d573fa) further. No need to use a `map` if you're not going to change/replace the contents in any way.